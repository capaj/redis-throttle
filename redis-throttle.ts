import { Redis } from 'ioredis'

const sleep = (ms: number) => {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export const redisThrottle = <T>({
  redis,
  key,
  limit,
  ttl,
  checkIntervalDivisor = 4,
  checkIntervalMilliseconds,
  fn
}: {
  redis: Redis
  key: string
  limit: number
  /**
   * the higher the divisor, the more accurate the throttling will be, but the more expensive it will be by using more redis.get calls, default is 4
   */
  checkIntervalDivisor?: number
  /**
   * optional, if not provided, it will be calculated from durationSeconds and checkIntervalDivisor
   */
  checkIntervalMilliseconds?: number
  ttl: number
  fn: (...args: any[]) => Promise<T>
}) => {
  let running: number
  let waiting = 0
  const getRunningRedisCount = async () => {
    running = parseInt((await redis.get(key)) ?? '0', 10)
    return running
  }

  const sleepMilliseconds =
    checkIntervalMilliseconds ?? (ttl / checkIntervalDivisor) * 1000
  const trampoline = async (
    args: any[],
    resolve: any,
    reject: any
  ): Promise<T | undefined> => {
    if (running && running > limit) {
      await sleep(sleepMilliseconds)
      return trampoline(args as [], resolve, reject) as Promise<T>
    }

    running = await getRunningRedisCount()

    if (running > limit) {
      await sleep(sleepMilliseconds)
      return trampoline(args as [], resolve, reject) as Promise<T>
    }
    const redisRes = await redis.multi().incr(key).expire(key, ttl).exec() // https://stackoverflow.com/a/50286772/671457 according to this multi is slowest. We could try to make it a bit faster

    // @ts-expect-error
    running = parseInt(redisRes[0][1], 10)
    if (running > limit) {
      running = await redis.decr(key)
      await sleep(sleepMilliseconds)
      return trampoline(args as [], resolve, reject) as Promise<T>
    }

    return fn(...(args as [])) as Promise<T>
  }

  return {
    waitingCount: () => waiting,
    runningCount: () => running ?? 0,
    runningRedisCount: getRunningRedisCount,
    reset: async () => {
      running = 0
      waiting = 0
      return redis.del(key)
    },
    call: async (...args: any[]): Promise<T> => {
      return new Promise((resolve, reject) => {
        waiting++
        trampoline(args as [], resolve, reject)
          .then(resolve as any)
          .catch(reject)
          .finally(async () => {
            running = await redis.decr(key)
            waiting = waiting - 1
          })
      })
    }
  }
}
