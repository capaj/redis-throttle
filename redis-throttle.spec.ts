import { describe, it, expect } from 'vitest'
import { redisThrottle } from './redis-throttle'
import Redis from 'ioredis'
import isCi from 'is-ci'

const redisClient = new Redis({
  host: isCi ? 'redis' : 'localhost'
})

const getTime = () => {
  {
    const hrTime = process.hrtime()
    return hrTime[0] * 1000 + hrTime[1] / 1000000
  }
}

describe('redis-throttle', () => {
  it('should resolve all promises without waiting', async () => {
    const throttled = redisThrottle({
      redis: redisClient,
      key: 'test',
      limit: 3,
      durationSeconds: 5,
      fn: async (n: number): Promise<number[]> => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve([n, throttled.runningCount(), throttled.waitingCount()])
          }, 500)
        })
      }
    })
    await throttled.reset()
    const before = getTime()
    const res = await Promise.all(
      [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map((n) => throttled.call(n))
    )
    expect(getTime() - before).toBeLessThan(5000)
    expect(getTime() - before).toBeGreaterThan(4000)
    expect(throttled.runningCount()).toBe(1)
    expect(throttled.waitingCount()).toBe(1)

    expect(res).toMatchSnapshot() // this can be flaky, but most of the time it passes ok
    expect(res.length).toBe(10)
    expect(await throttled.runningRedisCount()).toBe(0)
    expect(throttled.runningCount()).toBe(0)
    expect(throttled.waitingCount()).toBe(0)
  })
})
