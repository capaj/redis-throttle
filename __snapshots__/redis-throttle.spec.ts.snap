// Vitest Snapshot v1

exports[`redis-throttle > should resolve all promises without waiting 1`] = `
[
  [
    1,
    3,
    10,
  ],
  [
    2,
    3,
    10,
  ],
  [
    3,
    3,
    10,
  ],
  [
    4,
    3,
    7,
  ],
  [
    5,
    3,
    7,
  ],
  [
    6,
    3,
    7,
  ],
  [
    7,
    3,
    4,
  ],
  [
    8,
    3,
    4,
  ],
  [
    9,
    3,
    4,
  ],
  [
    10,
    1,
    1,
  ],
]
`;
